sky/core.py

```diff
@@ -437,26 +447,39 @@ def autostop(
         operation=operation,
     )
     backend = backend_utils.get_backend_from_handle(handle)
+
+    if tpu_utils.is_tpu_vm_pod(handle.launched_resources):
+        # Reference:
+        # https://cloud.google.com/tpu/docs/managing-tpus-tpu-vm#stopping_a_with_gcloud  # pylint: disable=line-too-long
+        raise exceptions.NotSupportedError(
+            f'{operation} cluster {cluster_name!r} with TPU VM Pod '
+            'is not supported.')
+
+    # Check cloud supports stopping spot instances
+    cloud = handle.launched_resources.cloud
+    assert cloud is not None, handle
+    supports_stop_spot = True
+    try:
+        cloud.check_features_are_supported(
+            {clouds.CloudImplementationFeatures.STOP_SPOT_INSTANCE})
+    except exceptions.NotSupportedError:
+        supports_stop_spot = False
+
     if not isinstance(backend, backends.CloudVmRayBackend):
         raise exceptions.NotSupportedError(
             f'{operation} cluster {cluster_name!r} with backend '
             f'{backend.__class__.__name__!r} is not supported.')
-    elif handle.launched_resources.use_spot and not down and not is_cancel:
+    elif (handle.launched_resources.use_spot and not down and not is_cancel and
+          not supports_stop_spot):
         # Disable spot instances to be autostopped.
-        # TODO(ewzeng): allow autostop for spot when stopping is supported.
+        #
+        # Exception: Allow GCP spot to be stopped since it preserves disk:
+        #   https://cloud.google.com/compute/docs/instances/preemptible#preemption-process  # pylint: disable=line-too-long
         raise exceptions.NotSupportedError(
             f'{colorama.Fore.YELLOW}Scheduling autostop on cluster '
             f'{cluster_name!r}...skipped.{colorama.Style.RESET_ALL}\n'
             '  Stopping spot instances is not supported as the attached '
             'disks will be lost.')
-
-    if tpu_utils.is_tpu_vm_pod(handle.launched_resources):
-        # Reference:
-        # https://cloud.google.com/tpu/docs/managing-tpus-tpu-vm#stopping_a_with_gcloud  # pylint: disable=line-too-long
-        raise exceptions.NotSupportedError(
-            f'{operation} cluster {cluster_name!r} with TPU VM Pod '
-            'is not supported.')
-
     # Check autostop is implemented for cloud
     cloud = handle.launched_resources.cloud
     if not down and idle_minutes >= 0:
```

Comment:
code gardening: Seems `CloudImplementationFeatures.AUTOSTOP` does not provide additional info than `CloudImplmentationFeatures.STOP`, maybe we should consider having them combined.

nit: Also, what do you think to have this line combined with L462-467?

Reply:
Good call, removed `CloudImplementationFeatures.AUTOSTOP`.

> nit: Also, what do you think to have this line combined with L462-467?

Hmm, I find it harder to reason since the call can raise. Separating STOP_SPOT_INSTANCE and STOP seems clearer?